### DISCUSSION TITLE
How do I verify a non-generic event was raised?

### DISCUSSION THREAD

**Discussion by tencek** [xunit/xunit#2772] [Question] [answered]
I know there's `public static RaisedEvent<T> Raises<T>(...)` in [EventAsserts.cs](https://github.com/xunit/assert.xunit/blob/b6cb339c16e3e190bc223d5324573a753e3b05bc/EventAsserts.cs#L32) but that only works for the `EventHandler<TEventArgs>`, AFAIK.
My event is of plain `EventHandler` type and I'd like to verify it was raised.
I ended up implementing it a bit dummy way on my own, here's the code:
```C#
    public class Testing
    {
        public static void RaisesEvent(
            Action<EventHandler> attach,
            Action<EventHandler> detach,
            Action testCode)
        {
            bool isEventRaised = false;
            var eventHandler = new EventHandler(delegate
            {
                isEventRaised = true;
            });

            attach(eventHandler);
            testCode();
            detach(eventHandler);

            Assert.True(isEventRaised);
        }
    }
```
Is there a way to achieve that using a standard xunit tools?
If not, would you consider adding it?

**Comment by bradwilson** [comment]
Yep, this is trivial for me to add.

Note: these will be `RaisesAny` and `RaisesAnyAsync` variants, not `Raises` and `RaisesAsync` because I will not be testing for exact types; anything which returns a value of `EventArgs` or anything derived from `EventArgs` will suffice. The reason is that trying to support an exact type match would yield the exact same signature as the existing `Assert.Raises<T>` method, since you'd need a generic type `T` to describe what the exact `EventArgs`-compatible type is that you want to test for.

Therefore, if you want to test for an exact type, you will need to run two assertions: `Assert.RaisesAny` to ensure the event was raised, followed by an `Assert.IsType<T>` against the returned `RaisedEvent.Arguments` property.

**Reply by bradwilson** [reply, marked as answer]
Fixed in #2773.

**Reply by tencek** [reply]
Thank you, that went fast! (-:
I'll give it a try.

**Reply by ChesireFreeThrow** [reply]
What about with events like in INotifyPropertyChanged using event PropertyChangedEventHandler? PropertyChanged; without a generic deinition? 

this works for me:
```
ISupportInitializeNotification sut = new TestService();

            var evt = Assert.RaisesAny(h => sut.Initialized += h, h => sut.Initialized -= h, () => sut.BeginInit());

            Assert.NotNull(evt);
            Assert.Equal(sut, evt.Sender);
            Assert.IsType<EventArgs>(evt.Arguments);
```

but not this:

```
var sut = new TestService();

            var evt = Assert.RaisesAny(h => sut.PropertyChanged += h, h => sut.PropertyChanged -= h, () => sut.BeginInit());

            Assert.NotNull(evt);
            Assert.Equal(sut, evt.Sender);
            Assert.IsType<PropertyChangedEventArgs>(evt.Arguments);
```

**Reply by bradwilson** [reply]
@ChesireFreeThrow Please open an issue with a full repro project. Thanks!
